static/components.ts

```diff
@@ -980,3 +938,275 @@ export function getDeviceViewWith(
+    }
+
+    if (!item.type) {
+        throw new Error(`Invalid ${location}: missing 'type' property`);
+    }
+
+    if (item.type === 'component') {
+        return validateComponentConfig(item, location);
+    }
+    if (item.type === 'row' || item.type === 'column' || item.type === 'stack') {
+        return validateLayoutItem(item, location);
+    }
+    throw new Error(`Invalid ${location}: unknown type '${item.type}'`);
+}
+
+/**
+ * Validates a component configuration
+ */
+function validateComponentConfig(config: any, location: string): ComponentConfig {
+    if (!config.componentName) {
+        throw new Error(`Invalid ${location}: missing 'componentName' property`);
+    }
+
+    if (typeof config.componentName !== 'string') {
+        throw new Error(`Invalid ${location}: 'componentName' must be a string`);
+    }
+
+    // Validate that the component state matches the expected type for this component
+    if (!validateComponentState(config.componentName, config.componentState)) {
+        throw new Error(
+            `Invalid ${location}: invalid component state for component '${config.componentName}'. ` +
+                `State: ${JSON.stringify(config.componentState, null, 2)}`,
+        );
+    }
+
+    return {
+        type: 'component',
+        componentName: config.componentName,
+        componentState: config.componentState,
+        title: config.title,
+        isClosable: config.isClosable,
+        reorderEnabled: config.reorderEnabled,
+        width: config.width,
+        height: config.height,
+    };
+}
+
+/**
+ * Validates a layout item (row, column, stack)
+ */
+function validateLayoutItem(item: any, location: string): LayoutItem {
+    if (!item.content || !Array.isArray(item.content)) {
+        throw new Error(`Invalid ${location}: layout items must have a 'content' array`);
+    }
+
+    return {
+        type: item.type as 'row' | 'column' | 'stack',
+        content: validateItemConfigs(item.content),
+        isClosable: item.isClosable,
+        reorderEnabled: item.reorderEnabled,
+        width: item.width,
+        height: item.height,
+        activeItemIndex: item.activeItemIndex,
+    };
+}
+
+/**
+ * Helper to convert to GoldenLayout's expected config format.
+ * This direction is safe since we're going from typed to untyped.
+ */
+export function toGoldenLayoutConfig(config: GoldenLayoutConfig): GoldenLayout.Config {
+    return config as GoldenLayout.Config;
+}
+
+/**
+ * Typed wrapper for createDragSource that avoids the need for 'as any'.
+ * Returns the result with _dragListener property for event handling.
+ *
+ * Note: We still need to cast internally because GoldenLayout's TypeScript
+ * definitions don't properly type the second parameter as accepting a function.
+ */
+export function createDragSource<K extends keyof ComponentStateMap>(
+    layout: GoldenLayout,
+    element: HTMLElement | JQuery,
+    factory: DragSourceFactory<K>,
+): any {
```

Comment:
Is golden layout missing a type for this?

Reply:
Will do

Reply:
Will check!

Reply:
There's no type BUT we came up with a slight improvement. PTAL